Travis docker

smiller171 opened the pull request:
@shazow For some reason the tests were hanging in Travis so I have them disabled in this PR (I don;t know enough about go to know why it's not working for me in Travis with tests, maybe you can help me figure that out.

This builds a Docker image that's only 3MB instead of 475MB. 

You need to set DOCKER_USER, DOCKER_EMAIL, and DOCKER_PASS environment variables in Travis for it to push the image to Docker Hub

shazow commented on Makefile:
```diff
@@ -7,7 +7,7 @@ SRCS = %.go
 
 $(BINARY): deps **/**/*.go **/*.go *.go
-	go build -ldflags "-X main.buildCommit `git describe --long --tags --dirty --always`" ./cmd/ssh-chat
+	CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -ldflags "-X main.buildCommit `git describe --long --tags --dirty --always`" ./cmd/ssh-chat
```
This will break for anyone who wants to run `make` on their own machine that is not linux.

smiller171 replied:
You're right. I should add an alternate script to build differently in travis

shazow replied:
How about just a standalone script for building the docker image that we can run from travis or locally?

shazow commented on .travis.yml:
```diff
@@ -9,9 +13,18 @@ install:
   - make deps
 
+go:
+  - 1.4
```
How about we bump it to 1.5 while we're here? (or do we need to specify? will it use the latest by default?)

smiller171 replied:
I think it will use the latest by default if you don't specify

shazow replied:
Let's keep it at the latest, I don't wanna be stuck on 1.4. :)

shazow commented on .travis.yml:
```diff
@@ -9,9 +13,18 @@ install:
+
 script:
   # TODO: Bring this back: - fgt golint
-  - make test
```
We use travis to run tests, I don't want to lose it.

smiller171 replied:
I agree, I want to fix this before merging

shazow commented:
Can we make this only run on the master branch after the PRs have been merged, in a separate suite?

I still want to keep travisci tests for PRs.

shazow commented:
> You need to set DOCKER_USER, DOCKER_EMAIL, and DOCKER_PASS environment variables in Travis for it to push the image to Docker Hub

Does it make any sense to do this? I don't want to let everyone in the world push to my docker hub account.

Maybe this shouldn't live in travis, but rather be a helper script I can run manually?

smiller171 commented:
@shazow I think I have it set up to only deploy to Docker hub on Master, not on PRs, but let me confirm that is working

smiller171 commented:
@shazow I fixed the issue with the makefile so that it's built differently in Travis for the purposes of packaging in Docker. I also changed the way the deployment is handled so that it will only deploy on master and tested that this works

shazow commented on docker-deploy.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash -ex
```
Can we include all the docker-specific build stuff here so that we don't break the Makefile for non-linux platforms?

smiller171 replied:
I already removed my changes to the makefile and am instead calling the Go command directly in the travis script.

smiller171 replied:
but of course that means make build is no longer being tested...

shazow replied:
Is that better than containing all docker-related things in here? I'd prefer to keep it in one place.

smiller171 replied:
I just updated it to build once with make build and then rebuild in the docker deploy script

shazow replied:
What's the benefit of having it build in the script step rather than in this bash script?

smiller171 replied:
It's just a better demarkation of where things happen. Build in the script and deploy in the deployment step.

That said in the context of the way you are running your tests it doesn't make a lot of difference, so I moved it into the deploy script.

I can't figure out why the tests are hanging though. As far as I can tell the only difference that could affect that is that it's now running on their legacy infrastructure (needed to build docker images for now) the tests also won't run for me locally.

shazow replied:
Sounds good, can you push that change? I'd really like to "contain" all the docker-related stuff in just one place. Especially since docker is rapidly changing/unstable and I don't want to go looking through several different files to maintain it. :)

smiller171 commented:
@shazow I'm getting the following when trying to run tests locally, and it's failing on Travis with my changes. I think the two may be related

```bash
make test
go test ./...
# ssh-chat
./auth_test.go:17: key.Public undefined (type *rsa.PrivateKey has no field or method Public)
./host_test.go:151: clientkey.Public undefined (type *rsa.PrivateKey has no field or method Public)
FAIL	ssh-chat [build failed]
ok  	ssh-chat/chat	0.002s
ok  	ssh-chat/chat/message	0.002s
?   	ssh-chat/cmd/ssh-chat	[no test files]
ok  	ssh-chat/sshd	0.039s
Makefile:30: recipe for target 'test' failed
make: *** [test] Error 2
```

smiller171 commented:
Ok, this is odd. I have confirmed that your tests run just fine in Travis' new Docker infrastructure, but completely hang in their legacy infrastructure.

They also won't run on my machine. Are you perhaps making some assumptions about the environment that don't work?

shazow commented:
What version of Go are you running? What OS?

smiller171 commented:
@shazow I'm on Ubuntu 15.04 with Go 1.4.

 Travis infrastructure is on Ubuntu 14.04, with Go 1.4 or 1.5

From a Travis perspective there is one version running on some kind of VM infrastructure and another running in Docker. The latter works, the first does not.

shazow commented:
Yea looks like the tests are hanging, probably a deadlock. Not sure why. We should add a `-timeout 10s` or something to the `go test` command. Default is 10min which is a bit cray.

smiller171 commented:
@shazow I don't really know go, but if you make the change I can merge it into this branch.

I wish I had any idea why it was hanging, but there should probably be a bug for tests not running properly in some environments

shazow commented:
If you can identify how you reproduce the failures on a specific platform, please open an issue. :)

smiller171 commented:
@shazow I have no idea how to create an environment where it actually runs. The only one I have access to is Travis' new Docker infrastructure

shazow commented:
Removed docker releases in favour of binary releases, can be easily wrapped with any blank docker now.
